### PR TITLE
feat: register contract route

### PR DESCRIPTION
**Pull request by asiaziola**

### CODE REVIEW

File: src/gateway/router/routes/registerRoute.ts [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    if (!isValid) {
+      throw new Error(`Data item binary is not valid.`);
+    }
+
+    const areContractTagsValid = await verifyContractTags(dataItem);
+    if (!areContractTagsValid) {
+      throw new Error(`Contract tags are not valid.`);
+    }
+
+    const bundlrResponse = await bundlr.uploader.transactionUploader(dataItem);
```
Comment (jakub-wojciechowski) [inline src/gateway/router/routes/registerRoute.ts:27]: Don't we check the bundlr response status?
Let's check if we received the full answer (we've got an issue when bundlr replied just 200, but without details and the transactions haven't been bundled)

Comment (jakub-wojciechowski) [reply src/gateway/router/routes/registerRoute.ts:27]: Let's at least check if the necessary fields from bundlr are returned (id, public)

Comment (asiaziola) [reply src/gateway/router/routes/registerRoute.ts:27]: you're right. proper check added.

File: src/gateway/router/routes/registerRoute.ts [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+      pst_ticker: type == 'pst' ? initState?.ticker : null,
+      pst_name: type == 'pst' ? initState?.name : null,
+      block_height: null,
+      block_timestamp: null,
+      content_type: contentType,
+      contract_tx: null,
+      bundler_contract_tx_id: bundlrResponse.data.id,
+      bundler_contract_node: BUNDLR_NODE2_URL,
+      bundler_contract_tags: JSON.stringify(dataItem.tags),
+      testnet: null,
```
Comment (jakub-wojciechowski) [inline src/gateway/router/routes/registerRoute.ts:66]: Let's have testnet as a parameter to the endpoint with default isTestnet = false

File: src/gateway/router/routes/registerRoute.ts [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+      owner: ownerAddress,
+      type: type,
+      pst_ticker: type == 'pst' ? initState?.ticker : null,
+      pst_name: type == 'pst' ? initState?.name : null,
+      block_height: null,
+      block_timestamp: null,
+      content_type: contentType,
+      contract_tx: null,
+      bundler_contract_tx_id: bundlrResponse.data.id,
+      bundler_contract_node: BUNDLR_NODE2_URL,
```
Comment (jakub-wojciechowski) [inline src/gateway/router/routes/registerRoute.ts:64]: It'll be better to extract it from bundlr rater than hardcoding (not critical)

File: src/gateway/router/routes/registerRoute.ts [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+      registeredContractId: bundlrResponse.data.id,
+    });
+
+    ctx.body = {
+      registeredContractId: bundlrResponse.data.id,
+    };
+  } catch (e) {
+    logger.error('Error while inserting data item.');
+    logger.error(e);
+    ctx.status = 500;
```
Comment (jakub-wojciechowski) [inline src/gateway/router/routes/registerRoute.ts:89]: Hmm, if we want to do it nicely errors like incorrect signature or missing tags should be returned as 400

**Review by jakub-wojciechowski — CHANGES_REQUESTED** [review]

File: src/gateway/router/routes/registerRoute.ts [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+    ctx.body = { message: e };
+  }
+}
+
+export async function verifyContractTags(dataItem: DataItem) {
+  const tags = dataItem.tags;
+  const tagsIncluded = [
+    new Tag('App-Name', 'SmartWeaveContract'),
+    new Tag('App-Version', '0.3.0'),
+    new Tag('Content-Type', 'application/x.arweave-manifest+json'),
```
Comment (jakub-wojciechowski) [inline src/gateway/router/routes/registerRoute.ts:99]: Can we safely limit the content to application/x.arweave-manifest+json. 
If so, we should return meaninful errors that will point to the wrong content-type

**Review by jakub-wojciechowski — CHANGES_REQUESTED** [review]

File: src/gateway/router/routes/registerRoute.ts (hunk 1 above)
Comment (ppedziwiatr) [reply src/gateway/router/routes/registerRoute.ts:27]: another issue here that  - in general - we're not storing in our db the response from bundlr from contracts deployments (only from interactions posted via sequencer).

It shouldn't be _too_ much work to add this to contract deployment (both the old way and the new, from this PR) - especially that probably we will be now spammed with thousands of contracts.

File: src/gateway/router/routes/registerRoute.ts [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+
+    logger.debug('Data item successfully bundled.', {
+      id: bundlrResponse.data.id,
+    });
+
+    const srcTxId = dataItem.tags.find((d) => d.name == 'Contract-Src')!.value;
+    const initStateRaw = dataItem.tags.find((d) => d.name == 'Init-State')!.value;
+    const initState = JSON.parse(initStateRaw);
+    const type = evalType(initState);
+    const ownerAddress = await arweave.wallets.ownerToAddress(bundlrResponse.data.public);
```
Comment (ppedziwiatr) [inline src/gateway/router/routes/registerRoute.ts:48]: Noooo!!! `bundlrResponse.data.public` is the jwk.n of the Bundlr node, not the jwk.n of the signed dataitem (and I guess we want to set here the DataItem's signer jwk.n, no the Bundlr's one)

File: tools/alex.ts [hunk 7]
```diff
@@ -0,0 +1,77 @@
+/* eslint-disable */
+import fs from 'fs';
+import Bundlr from '@bundlr-network/client';
+import { LoggerFactory, WarpFactory, defaultCacheOptions } from 'warp-contracts';
+import { ArweaveSigner } from 'arbundles/src/signing';
+import { createData } from 'arbundles';
+
+async function main() {
```
Comment (ppedziwiatr) [inline tools/alex.ts:8]: why is the Alex project named 'alex'? 🤡

File: src/gateway/router/routes/registerRoute.ts (hunk 3 above)
Comment (ppedziwiatr) [reply src/gateway/router/routes/registerRoute.ts]: Not from the response, but from the bundlr object (should be available under ctx.bundlr, and then bundlr.api (similar to arweave.api))

But this should be probably refactored in all places, not sure if that is a critical "issue" now ;-)

Comment (asiaziola) [reply src/gateway/router/routes/registerRoute.ts]: ah, ok. let's leave it for now, I don't want this pr to be too dangerous to install :))))

File: src/gateway/router/routes/registerRoute.ts (hunk 5 above)
Comment (asiaziola) [reply src/gateway/router/routes/registerRoute.ts]: yes, done

**Comment by ppedziwiatr** [comment]
YOLO 🙈

**Pull request merged**